File: Sources/CreatorCoreForge/Ebook2AudiobookBridge.swift
import Foundation

/// Bridge to run the Python-based `ebook2audiobook` snapshot from Swift.
/// This allows any CoreForge app to invoke the advanced conversion pipeline
/// packaged under `apps/ebook2audiobook`.
public struct Ebook2AudiobookBridge {
    public var scriptPath: String

    /// Initialize the bridge with an optional custom script path.
    public init(scriptPath: String = "apps/ebook2audiobook/app.py") {
        self.scriptPath = scriptPath
    }

    /// Convert the given ebook file to an audiobook using the Python script.
    /// - Parameters:
    ///   - ebook: Path to the input ebook file.
    ///   - outputDirectory: Where the generated audio files will be stored.
    /// - Returns: `true` if the conversion succeeded.
    @discardableResult
    public func convert(ebook: String, outputDirectory: String) -> Bool {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = ["python3", scriptPath, "--headless", "--ebook", ebook, "--output_dir", outputDirectory]

        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = pipe

        do {
            try process.run()
            process.waitUntilExit()
            return process.terminationStatus == 0
        } catch {
            print("Failed to launch ebook2audiobook: \(error)")
            return false
        }
    }
}

File: scripts/ebook2audiobook_bridge.py
#!/usr/bin/env python3
"""Run the ebook2audiobook snapshot to convert an ebook to audio."""
import argparse
import os
import subprocess
import sys


def main():
    parser = argparse.ArgumentParser(
        description="Convert an ebook to audio using the bundled ebook2audiobook script")
    parser.add_argument("ebook", help="Path to the ebook file")
    parser.add_argument("-o", "--output", default="output", help="Directory for generated audio")
    args = parser.parse_args()

    script_path = os.path.join("apps", "ebook2audiobook", "app.py")
    cmd = [sys.executable, script_path, "--headless", "--ebook", os.path.abspath(args.ebook),
           "--output_dir", os.path.abspath(args.output)]
    subprocess.run(cmd, check=True)

if __name__ == "__main__":
    main()
